presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDirectStreamReader.java

```diff
@@ -144,47 +149,69 @@ public Block readBlock(Type type)
             }
         }
 
+        offsetVector[0] = 0;
         long totalLength = 0;
         for (int i = 0; i < nextBatchSize; i++) {
+            offsetVector[i + 1] = offsetVector[i];
             if (!isNullVector[i]) {
                 totalLength += lengthVector[i];
+                offsetVector[i + 1] += lengthVector[i];
             }
         }
+
+        int currentBatchSize = nextBatchSize;
+        readOffset = 0;
+        nextBatchSize = 0;
+        if (totalLength == 0) {
+            return new VariableWidthBlock(currentBatchSize, EMPTY_SLICE, offsetVector, isNullVector);
+        }
         if (totalLength > ONE_GIGABYTE) {
             throw new PrestoException(GENERIC_INTERNAL_ERROR,
                     format("Values in column \"%s\" are too large to process for Presto. %s column values are larger than 1GB [%s]", streamDescriptor.getFieldName(), nextBatchSize, streamDescriptor.getOrcDataSourceId()));
         }
-
-        byte[] data = EMPTY_BYTE_ARRAY;
-        if (totalLength > 0) {
-            if (dataStream == null) {
-                throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
-            }
-            data = dataStream.next(toIntExact(totalLength));
+        if (dataStream == null) {
+            throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
         }
 
-        Slice[] sliceVector = new Slice[nextBatchSize];
-
-        int offset = 0;
-        for (int i = 0; i < nextBatchSize; i++) {
-            if (!isNullVector[i]) {
-                int length = lengthVector[i];
-                Slice value = Slices.wrappedBuffer(data, offset, length);
-                if (isVarcharType(type)) {
-                    value = truncateToLength(value, type);
-                }
-                if (isCharType(type)) {
-                    value = trimSpacesAndTruncateToLength(value, type);
-                }
-                sliceVector[i] = value;
-                offset += length;
+        // allocate enough space to read
+        byte[] data = new byte[toIntExact(totalLength)];
+        Slice slice = Slices.wrappedBuffer(data);
+
+        // truncate string and update offsets
+        for (int i = 0; i < currentBatchSize; i++) {
+            // the previously computed offsets are original offsets without truncation
+            // update the offsets in this loop to be the truncated ones
+            offsetVector[i + 1] = offsetVector[i];
+            if (isNullVector[i]) {
+                continue;
+            }
+            int offset = offsetVector[i];
+            int length = lengthVector[i];
+
+            // read data without truncation
+            dataStream.next(offset + length, data, offset);
+
+            // calculate truncated length
+            int truncatedLength;
```

Comment:
How do you feel about rewriting the below code like this? I got rid of the comment in the first `if` and got rid of the final `else`, also added new `codePointCount` local variables, because at first look it is confusing to see that we are passing some `length`s to `getTruncatedLength` as the third and fourth arguments, and I wanted to make it a bit more explicit that what we are passing as the last argument is really the number of code points.

```
// calculate truncated length
int truncatedLength = length;
if (isVarcharType(type)) {
    VarcharType varcharType = (VarcharType) type;
    int codePointCount = varcharType.isUnbounded() ? length : varcharType.getLengthSafe();
    truncatedLength = getTruncatedLength(slice, offset, length, codePointCount);
}
else if (isCharType(type)) {
    int trimmedLength = sliceLengthWithoutTrailingSpaces(slice, offset, length);
    verify(trimmedLength >= 0);
    int codePointCount = ((CharType) type).getLength();
    truncatedLength = getTruncatedLength(slice, offset, trimmedLength, codePointCount);
}
```